inference/aot_serialization.md

```diff
@@ -0,0 +1,236 @@
+        "timestep": torch.load("timestep.pt", map_location="cuda") / 1000,
+        "img_ids": torch.load("latent_image_ids.pt", map_location="cuda"),
+        "txt_ids": torch.load("text_ids.pt", map_location="cuda"),
+        "guidance": torch.load("guidance.pt", map_location="cuda"),
+        "joint_attention_kwargs": None, 
+        "return_dict": False
+    }
+    return example_inputs
+
+def benchmark_fn(f, *args, **kwargs):
+    t0 = benchmark.Timer(
+        stmt="f(*args, **kwargs)",
+        globals={"args": args, "kwargs": kwargs, "f": f},
+        num_threads=torch.get_num_threads(),
+    )
+    return f"{(t0.blocked_autorange().mean):.3f}"
+
+def load_model():
+    model = FluxTransformer2DModel.from_pretrained(
+        "black-forest-labs/FLUX.1-dev", subfolder="transformer", torch_dtype=torch.bfloat16
+    ).to("cuda")
+    return model
+
+def aot_compile(name, fn, **sample_kwargs):
+    path = f"{name}.so"
+    print(f"{path=}")
+    options = {
+        "aot_inductor.output_path": path,
+        "max_autotune": True,
+        "triton.cudagraphs": True,
+    }
+
+    torch._export.aot_compile(
+        fn,
+        (),
+        sample_kwargs,
+        options=options,
+        disable_constraint_solver=True,
+    )
+    return path 
+
+def aot_load(path):
+    return torch._export.aot_load(path, "cuda")
+
+@torch.no_grad()
+def f(model, **kwargs):
+    return model(**kwargs)
+
+model = load_model()
+num_channels_latents = model.config.in_channels // 4 
+
+inputs = get_example_inputs()
+path = aot_compile("bs_1_1024", partial(f, model=model), **inputs)
+
+compiled_func_1 = aot_load(path)
+print(f"{compiled_func_1(**inputs)[0].shape=}")
+
+for _ in range(5):
+    _ = compiled_func_1(**inputs)[0]
+
+time = benchmark_fn(f, compiled_func_1, **inputs)
+print(time) # 0.421 seconds on an A100.
+```
+
+</details>    
+
+Here, we’re obtaining the compiled graph for a batch size of 1 and a resolution of 1024x1024. The example inputs for the AoT compilation were manually obtained from `pipeline_flux.py` ([implementation](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux.py)) by serializing the inputs (inputs that go to `self.transformer` [here](https://github.com/huggingface/diffusers/blob/fddbab79932eedf1a78041ef38c47df80ab84c90/src/diffusers/pipelines/flux/pipeline_flux.py#L732)) to have reasonable values rather than random values. 
+
+Then, run the full pipeline:
+
+```python
+import torch 
+from diffusers import DiffusionPipeline
+
+pipeline = DiffusionPipeline.from_pretrained(
+    "black-forest-labs/FLUX.1-dev",
+    transformer=None,
+    torch_dtype=torch.bfloat16,
+).to("cuda")
+pipeline.transformer = torch._export.aot_load("/raid/.cache/bs_1_1024.so", "cuda")
+
+image = pipeline(
+    "cute dog", guidance_scale=3.5, max_sequence_length=512, num_inference_steps=50
+).images[0]
+image.save("aot_compiled.png")
+```
+
+Note that running the above requires the following patch:
+
+```diff
```

Comment:
why is this needed? is this something specific to flux or some bug in aot_compile/load?

Reply:
Specific the AoT compiled function.

Reply:
I'm trying to understand how people will use this.

Is this tutorial going to be used by other model code as well? if so, how do they know how to make similar changes to their model

Reply:
They will apply similar changes to their pipeline code to make it compatible.

Reply:
I see, that sounds good, I think it's fine